clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -451,68 +487,71 @@ public int compare(ProducerBatch o1, ProducerBatch o2) {
-        if (queue == null)
-            return;
-        queue.remove(batch);
+        if (hasInflightBatches(batch.topicPartition)) {
+            PriorityQueue<ProducerBatch> queue = topicPartitionBookkeeper.getPartition(batch.topicPartition).inflightBatchesBySequence;
+            queue.remove(batch);
+        }
     }
 
     synchronized void maybeUpdateLastAckedSequence(TopicPartition topicPartition, int sequence) {
-        if (sequence > lastAckedSequence(topicPartition))
-            lastAckedSequence.put(topicPartition, sequence);
+        if (sequence > lastAckedSequence(topicPartition).orElse(NO_LAST_ACKED_SEQUENCE_NUMBER))
+            topicPartitionBookkeeper.getPartition(topicPartition).lastAckedSequence = sequence;
     }
 
-    synchronized int lastAckedSequence(TopicPartition topicPartition) {
-        Integer currentLastAckedSequence = lastAckedSequence.get(topicPartition);
-        if (currentLastAckedSequence == null)
-            return -1;
-        return currentLastAckedSequence;
+    synchronized OptionalInt lastAckedSequence(TopicPartition topicPartition) {
+        return topicPartitionBookkeeper.lastAckedSequence(topicPartition);
     }
 
-    synchronized long lastAckedOffset(TopicPartition topicPartition) {
-        Long offset = lastAckedOffset.get(topicPartition);
-        if (offset == null)
-            return ProduceResponse.INVALID_OFFSET;
-        return offset;
+    synchronized OptionalLong lastAckedOffset(TopicPartition topicPartition) {
+        return topicPartitionBookkeeper.lastAckedOffset(topicPartition);
     }
 
     synchronized void updateLastAckedOffset(ProduceResponse.PartitionResponse response, ProducerBatch batch) {
         if (response.baseOffset == ProduceResponse.INVALID_OFFSET)
             return;
         long lastOffset = response.baseOffset + batch.recordCount - 1;
-        if (lastOffset > lastAckedOffset(batch.topicPartition)) {
-            lastAckedOffset.put(batch.topicPartition, lastOffset);
+        OptionalLong lastAckedOffset = lastAckedOffset(batch.topicPartition);
+        // It might happen that the TransactionManager has been reset while a request was reenqueued and got a valid
+        // response for this. This can happen only if the producer is only idempotent (not transactional) and in
+        // this case there will be no tracked bookkeeper entry about it, so we have to insert one.
+        if (!lastAckedOffset.isPresent() && !isTransactional()) {
+            ensureHasBookkeeperEntry(batch.topicPartition);
+        }
+        if (lastOffset > lastAckedOffset.orElse(ProduceResponse.INVALID_OFFSET)) {
+            topicPartitionBookkeeper.getPartition(batch.topicPartition).lastAckedOffset = lastOffset;
         } else {
             log.trace("Partition {} keeps lastOffset at {}", batch.topicPartition, lastOffset);
         }
     }
 
+    private void ensureHasBookkeeperEntry(TopicPartition topicPartition) {
```

Comment:
nit: seems we don't need this method anymore since `addPartitions` is idempotent.